File: Data Structures & Algorithms/count-number-of-islands/submission-3.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        if (grid == null || grid.Length == 0) return 0;

        int ROWS = grid.Length;
        int COLS = grid[0].Length;
        HashSet<(int r, int c)> visitedLand = new HashSet<(int r, int c)>();
        int islandCount = 0;

        void Dfs(int r, int c) {
            if (Math.Min(r, c) < 0 || 
                r == ROWS || 
                c == COLS ||
                grid[r][c] == '0' || 
                visitedLand.Contains((r, c))) return;
            
            visitedLand.Add((r, c));
            Dfs(r - 1, c);
            Dfs(r, c - 1);
            Dfs(r + 1, c);
            Dfs(r, c + 1);
        }

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == '1' && !visitedLand.Contains((r, c))) {
                    Dfs(r, c);
                    islandCount++;
                }
            }
        }

        return islandCount;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-3.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        visited_land = set()
        max_area = 0

        def dfs(r, c):
            if (min(r, c) < 0 or 
                r == ROWS or 
                c == COLS or 
                grid[r][c] == 0 or 
                (r, c) in visited_land):
                return 0
            
            visited_land.add((r, c))
            curr_area = 1
            curr_area += dfs(r - 1, c)
            curr_area += dfs(r, c + 1)
            curr_area += dfs(r + 1, c)
            curr_area += dfs(r, c - 1)

            return curr_area

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1 and (r, c) not in visited_land:
                    max_area = max(max_area, dfs(r, c))
        
        return max_area